Add additional property support by subscripting model

packyg opened the pull request:
Implements #218

Add support for `additionalProperties` by holding extra props in a private dict.
Can be accessed with pass-through subscripting dunder methods, including a `@property` to get the keys (see the bottom of `model.pyi`)


- Updated `from_dict` to `pop` all named properties (from a shallow copy), with the rest going in the catch-all
  - Required saving the value to a temp variable for inspection
- `to_dict` sets the additional properties first, giving precedent to named props


This would be a big win for my team - we had been relying on the previous behavior of having dicts when `type=object` to handle cases that we haven't or can't model very well.

packyg commented:
@dbanty I'm interested to see if you like this approach. If you do, I'll go ahead and update the tests.

(cc @bowenwr @dtkav)

dbanty commented on openapi_python_client/templates/model.pyi:
```diff
@@ -33,14 +46,17 @@ class {{ model.reference.class_name }}:
             {% endif %}
             {% endfor %}
-        }
+        })
```
Why not leave the regular properties in the field dict declaration? Would be a bit faster I think rather than having to update after the fact, even without having additional properties.

dbanty commented on openapi_python_client/templates/model.pyi:
```diff
@@ -66,8 +83,37 @@ class {{ model.reference.class_name }}:
         )
+
+        {% if model.additional_properties %}
+        {{model.reference.module_name}}._additional_properties.update(d)
```
Why update vs set directly? I would think assignment would be faster but maybe not.

dbanty commented on openapi_python_client/templates/model.pyi:
```diff
@@ -66,8 +83,37 @@ class {{ model.reference.class_name }}:
+
+    @property
+    def additional_properties(self) -> Dict[str, {{ additional_property_type }}]:
+        return self._additional_properties
```
What's the benefit of having this internal w/ `@property` vs just declaring it as `additional_properties` originally?

dbanty reviewed (changes requested):
Overall I think it looks good. I'm a little uneasy about putting set/getitem directly on the model but I don't have a good reason for being uneasy so it's probably fine 😅. I pointed out a couple places I think we could clean up some of the dict changes. Other than that, I would like @emannguitar to take a look at this as well to get his opinion.

emannguitar reviewed (commented):
LGTM outside of @dbanty's comments

emannguitar commented:
Actually two small things: 
1. Per the OpenAPI spec, `type: object` by itself (i.e. without specifying `additionalProperties: true` or `additionalProperties: {}`) is also a free-form object
1. Given that a free-form object is really just an arbitrary dict at the end of the day, they should just be dicts instead of a model with nothing but `_aditional_properties`

packyg reviewed (commented):
Thanks for the feedback @dbanty and @emannguitar.

I realized I also didn't properly handle serialization/deserialization when the schema of the `additionalProperties` is specified and the dict holds models. I will take a crack at that.

> 1. Per the OpenAPI spec, type: object by itself (i.e. without specifying additionalProperties: true or additionalProperties: {}) is also a free-form object

This wasn't immediately clear to me. I know this is the case for JSON schema, wasn't sure it was for OpenAPI, but it appears so "_Consistent with JSON Schema, `additionalProperties` defaults to true._" 
I will revisit to account for this.

-----

> 2. Given that a free-form object is really just an arbitrary dict at the end of the day, they should just be dicts instead of a model with nothing but _aditional_properties

This is true and was my initial thought for objects with `additionalProperties` and no named `properties`, but it gets more complicated when we get to the case I mentioned at the top of this comment and it's a `Dict[str, SomeModelClass]`

packyg commented:
OK, I've now updated this to properly interpret `additionalProperties` per the spec (defaulting to true)

I also added support for `additionalProperties` where the elements are modeled. This requires calling `to_dict`/`from_dict` for serialization/deserialization, so we can't just use a raw `dict` there.

packyg commented on openapi_python_client/templates/property_templates/model_property.pyi:
```diff
@@ -2,9 +2,16 @@
 {% else %}
+{% if initial_value != None %}
 {{ property.python_name }} = {{ initial_value }}
-if {{ source }} is not None:
-    {{ property.python_name }} = {{ property.reference.class_name }}.from_dict(cast(Dict[str, Any], {{ source }}))
+{% elif property.nullable %}
+{{ property.python_name }} = None
+{% else %}
+{{ property.python_name }} = UNSET
+{% endif %}
```
This addresses what I think was a bug before

packyg commented on openapi_python_client/templates/property_templates/model_property.pyi:
```diff
@@ -16,7 +23,7 @@ if {{ source }} is not None:
 {% endif %}
 {% else %}
-{{ destination }}{% if declare_type %}: {{ property.get_type_string() }}{% endif %} = UNSET
+{{ destination }}{% if declare_type %}: Union[{% if property.nullable %}None, {% endif %}Unset, Dict[str, Any]]{% endif %} = UNSET
```
Same here - the type was wrong here

emannguitar reviewed (approved):
LGTM!

@dbanty I generated a client for inventory to test and this seems to fix the issues we were having

dbanty reviewed (approved):
Awesome, fantastic work! I'll put together a release including this feature shortly.
